### PR TITLE
Cst-5673 implement patch operations for action and policyType attributes

### LINKED ISSUE #7812: [DS-4477] Implement Resource Policy patch operations for policyType and action properties
<i>Imported from JIRA <a href="https://jira.lyrasis.org/browse/DS-4477">[DS-4477]</a> created by giuseppe.digilio</i>
<p>Replace operation for properties policyType and action are not supported by resourcepolicies endpoint. Implement them</p>

### PR DESCRIPTION
**Pull request by ballarinie**
## References
* Fixes #7812 

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 1]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
                 hasJsonPath("$.name", is(name)))));
     }
 
+    @Test
+    public void patchAddActionTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2209]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 2]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$", Matchers.allOf(
+                hasJsonPath("$.action", is(Constants.actionText[action])))));
+    }
+
+    @Test
+    public void patchReplaceActionTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2257]: as above

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 3]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .content(patchBody)
+            .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+            .andExpect(status().isNotFound());
+    }
+
+    @Test
+    public void patchReplaceActionUnprocessableEntityTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2397]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 4]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .content(patchBody)
+            .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+            .andExpect(status().isUnprocessableEntity());
+    }
+
+    @Test
+    public void patchAddPolicyTypeTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2438]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 5]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .andExpect(jsonPath("$", Matchers.allOf(
+                hasJsonPath("$.action", is(Constants.actionText[resourcePolicy.getAction()])),
+                hasJsonPath("$.policyType", is(policyType)))));
+    }
+
+    @Test
+    public void patchRemovePolicyTypeTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2488]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 6]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .andExpect(jsonPath("$", Matchers.allOf(
+                hasJsonPath("$.action", is(Constants.actionText[resourcePolicy.getAction()])))))
+            .andExpect(jsonPath("$.policyType").doesNotExist());
+    }
+
+    @Test
+    public void patchReplacePolicyTypeTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2538]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 7]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .andExpect(jsonPath("$", Matchers.allOf(
+                hasJsonPath("$.action", is(Constants.actionText[resourcePolicy.getAction()])),
+                hasJsonPath("$.policyType", is(resourcePolicy.getRpType())))));
+    }
+
+    @Test
+    public void patchReplacePolicyTypeForbiddenTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2627]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java [hunk 8]
```diff
@@ -2202,6 +2202,518 @@ public void patchAddNameTest() throws Exception {
...
+            .content(patchBody)
+            .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+            .andExpect(status().isNotFound());
+    }
+
+    @Test
+    public void patchReplacePolicyTypeBadRequestTest() throws Exception {
+        context.turnOffAuthorisationSystem();
+
+        EPerson eperson1 = EPersonBuilder.createEPerson(context)
```
Comment (corrad82-4s) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/ResourcePolicyRestRepositoryIT.java:2680]: Here already defined user in field "eperson" of base class can be used without creating a new one

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/resourcePolicy/ResourcePolicyActionAddOrReplaceOperation.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Performs the actual add or replace action of resourcePolicy operation. Both
+     * actions are allowed since the starting value of action is a defined int.
+     * 
+     * @param resourcePolicy resourcePolicy being patched
+     * @param operation      patch operation
+     */
+    private void replace(ResourcePolicy resourcePolicy, Operation operation) {
+        int action = (int) operation.getValue();
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/resourcePolicy/ResourcePolicyActionAddOrReplaceOperation.java:56]: This line throws a `ClassCastException` for me anytime I try to change the `action` of an existing Policy. 

For example, I tried to edit a Policy which specified `READ` and change it to `WRITE`.   I received a 500 response, and got this error in the backend logs:
```
2022-05-18 19:40:18,210 ERROR a14cfb2c-4e21-445c-bdb9-b1f53afe8843 475691ec-6638-4c9b-b96a-7e7cf7572654 org.dspace.app.r
est.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:500)
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Integer (java.lang.String and jav
a.lang.Integer are in module java.base of loader 'bootstrap')
        at org.dspace.app.rest.repository.patch.operation.resourcePolicy.ResourcePolicyActionAddOrReplaceOperation.repla
ce(ResourcePolicyActionAddOrReplaceOperation.java:56) ~[classes/:7.3-SNAPSHOT]
        at org.dspace.app.rest.repository.patch.operation.resourcePolicy.ResourcePolicyActionAddOrReplaceOperation.perfo
rm(ResourcePolicyActionAddOrReplaceOperation.java:41) ~[classes/:7.3-SNAPSHOT]
        at org.dspace.app.rest.repository.patch.ResourcePatch.performPatchOperation(ResourcePatch.java:58) ~[classes/:7.
3-SNAPSHOT]
        at org.dspace.app.rest.repository.patch.ResourcePatch.patch(ResourcePatch.java:42) ~[classes/:7.3-SNAPSHOT]
        at org.dspace.app.rest.repository.ResourcePolicyRestRepository.patch(ResourcePolicyRestRepository.java:320) ~[cl
asses/:7.3-SNAPSHOT]
```

**Review by tdonohue — CHANGES_REQUESTED** [review]
@ballarinie : This isn't fully working for me when I try to test it with the UI PR at https://github.com/DSpace/dspace-angular/pull/1641

From the UI, I *can* change the `policyType` successfully of an existing Policy. But, I *cannot* change the `action`.  It initially "appears" to work, but the backend throws a 500 error & if you reload the UI page, the change is gone.  I noted the line which errors out below.  To reproduce this, here's all I did:

* Login as an Admin
* Edit an Item's & view authorization policies
* Attempt to change Item `action=READ` policy to be `action=WRITE`
* UI will show success. But, if you look at the backend logs, you'll see a 500 error. 
* Reload page in UI, and the policy will revert to `action=READ`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/resourcePolicy/ResourcePolicyActionAddOrReplaceOperation.java (hunk 9 above)
Comment (ballarinie) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/resourcePolicy/ResourcePolicyActionAddOrReplaceOperation.java:56]: Thanks for the report, we'll take it in charge and work on it in the next few days.

**Review by tdonohue — APPROVED** [review]
👍 @ballarinie : Thanks for the updates!  This is now working properly.

I think it's nearly ready to merge, however it looks like @corrad82-4s 's feedback hasn't been addressed.  As he noted, in your ITs, you can remove all the code which is creating a new Eperson using `EPersonBuilder`.  Instead, you should just use the pre-defined `eperson` object which is created for all ITs in the `AbstractIntegrationTestWithDatabase` here: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/test/java/org/dspace/AbstractIntegrationTestWithDatabase.java#L60

Once that minor cleanup is done, I think this is ready to be merged.  Thanks again!

**Comment by ballarinie** [comment]
I apologize for the delay. I have now followed @corrad82-4s 's suggestions and solved the conflicts.

**Review by tdonohue — APPROVED** [review]
Thanks @ballarinie !  I've verified that @corrad82-4s 's prior feedback has been addressed.

**Pull request merged**